src/azul/bigquery_reservation.py

```diff
@@ -78,53 +91,80 @@ def _reservation_parent_path(self) -> str:
+        self._refresh('capacity_commitment')
+        if self.capacity_commitment_name is not None:
+            log.info('Slot commitment already purchased')
         else:
-            self.capacity_commitment_name = self._purchase_commitment().name
-            self.reservation_name = self._create_reservation().name
-            self.assignment_name = self._create_assignment(self.reservation_name).name
+            self._purchase_commitment()
 
-    def ensure_slots_deleted(self) -> None:
+    def create_reservation(self) -> None:
+        """
+        Idempotently create reservation.
+        """
+        self._refresh('reservation')
+        if self.reservation_name is not None:
+            log.info('Reservation already created')
+        else:
+            self._create_reservation()
+
+    def assign_slots(self) -> None:
         """
-        Idempotently delete flex slots.
+        Idempotently assign capacity commitment to a reservation.
         """
-        if not self.has_active_slots():
-            log.info('No slot commitment active')
-        elif self.dry_run:
-            log.info('No resources deleted because dry_run was set to True.')
+        self._refresh('assignment')
+        if self.assignment_name is not None:
+            log.info('Slots already assigned')
         else:
-            for resource_type in ('assignment',
-                                  'reservation',
-                                  'capacity_commitment'):
-                attr_name = resource_type + '_name'
-                resource_name = getattr(self, attr_name)
-                delete_method = getattr(self._client, 'delete_' + resource_type)
-                delete_method(name=resource_name)
-                log.info('Deleted resource %r', f'{resource_type}:{resource_name}')
-                setattr(self, attr_name, None)
+            self._create_assignment(self.reservation_name)
+
+    def ensure_slots_deleted(self) -> None:
+        """
+        Idempotently delete all resources.
+        """
+        for resource_type in ('assignment',
+                              'reservation',
+                              'capacity_commitment'):
+            attr_name = resource_type + '_name'
+            resource_name = getattr(self, attr_name)
+            if resource_name is None:
+                log.info('%r does not exist', resource_type)
+            else:
+                resource_str = f'{resource_type}:{resource_name}'
+                if self.dry_run:
+                    log.info('Did not delete resource %r because dry_run was set to True',
+                             resource_str)
+                else:
+                    delete_method = getattr(self._client, 'delete_' + resource_type)
+                    delete_method(name=resource_name)
+                    log.info('Deleted resource %r', resource_str)
+        self.refresh()
+        require(self.dry_run or self.state() is SlotState.inactive,
```

Comment:
`require` is for checking conditions on input. `assert` is for checking invariants. I don't think this is either.

Here and elsewhere.